java-checks/src/main/java/org/sonar/java/checks/tests/AssertJChainSimplificationCheck.java

```diff
@@ -0,0 +1,132 @@
+public class AssertJChainSimplificationCheck extends AbstractMethodDetection {
+  private static final String ABSTRACT_ASSERT = "org.assertj.core.api.AbstractAssert";
+  private static final List<String> ASSERTION_MESSAGE_NAMES = Arrays.asList("as", "describedAs", "withFailMessage", "overridingErrorMessage");
+
+  private static final MethodMatchers ASSERTION_PREDICATES = MethodMatchers.create().ofSubTypes(ABSTRACT_ASSERT)
+    .name(name -> !ASSERTION_MESSAGE_NAMES.contains(name)).withAnyParameters().build();
+
+  private static final MethodMatchers ASSERTIONS_SUBJECT_METHODS = MethodMatchers.create().ofTypes(
+    "org.assertj.core.api.Assertions",
+    "org.assertj.core.api.AssertionsForInterfaceTypes",
+    "org.assertj.core.api.AssertionsForClassTypes")
+    .names("assertThat", "assertThatObject").withAnyParameters().build();
+
+  /**
+   * @see AssertJChainSimplificationIndex#CONTEXT_FREE_SIMPLIFIERS
+   */
+  private static final Map<String, List<SimplifierWithoutContext>> CONTEXT_FREE_SIMPLIFIERS = AssertJChainSimplificationIndex.CONTEXT_FREE_SIMPLIFIERS;
+
+  /**
+   * @see AssertJChainSimplificationIndex#SIMPLIFIERS_WITH_CONTEXT
+   */
+  private static final Map<String, List<SimplifierWithContext>> SIMPLIFIERS_WITH_CONTEXT = AssertJChainSimplificationIndex.SIMPLIFIERS_WITH_CONTEXT;
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return ASSERTIONS_SUBJECT_METHODS;
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree subjectMit) {
+    List<MethodInvocationTree> predicates = new ArrayList<>();
+    Optional<MethodInvocationTree> nextPredicate = MethodTreeUtils.subsequentMethodInvocation(subjectMit, ASSERTION_PREDICATES);
+
+    while (nextPredicate.isPresent()) {
+      predicates.add(nextPredicate.get());
+      nextPredicate = MethodTreeUtils.subsequentMethodInvocation(nextPredicate.get(), ASSERTION_PREDICATES);
+    }
+
+    boolean wasIssueRaised = checkPredicatesForSimplification(
+      predicates, CONTEXT_FREE_SIMPLIFIERS, SimplifierWithoutContext::simplify,
+      (predicate, replacement) -> reportIssue(ExpressionUtils.methodName(predicate),
+        String.format("Use %s instead", replacement)));
+
+    // We do not continue when we have already raised an issue to avoid potentially conflicting issue reports. If we
+    // have more than one predicate we also avoid continuing to avoid FP on cases such as:
+    // assertThat(Integer.valueOf(1).compareTo(2)).isGreaterThan(1).isLessThan(10)
+    if (wasIssueRaised || predicates.size() > 1) {
+      return;
+    }
+
+    checkPredicatesForSimplification(
+      predicates, SIMPLIFIERS_WITH_CONTEXT, (simplifier, predicate) -> simplifier.simplify(subjectMit, predicate),
+      (predicate, replacement) -> reportIssue(ExpressionUtils.methodName(predicate),
+        String.format("Use %s instead", replacement),
+        Collections.singletonList(
+          new JavaFileScannerContext.Location("", subjectMit)),
```

Comment:
When the secondary location message is empty, SonarQube seems buggy, it's not possible to navigate to a secondary location. Here, it's the "assertion subject". Or we can probably highlight only the `subjectMit`'s first argument as "actual value".
Or we can also completely remove the secondary location so we can hard code `reportIssue` in `checkPredicatesForSimplification` and get rid of the `BiConsumer<MethodInvocationTree, String> reportingMethod`

Reply:
What was our final agreement in regards to the message here? Is it fine to leave the message blank? (I think that makes sense, although we should investigate the SQ issue at the same time).